### PR TITLE
Version code generation based on tag name

### LINKED ISSUE #6086: Improve versionCode generation in build.cake for AppStore/PlayStore builds
## The Problem
We are currently incrementing version code based on the total number of commits for our AppStore/PlayStore builds (`rev-list --count HEAD`). This can lead to issues since we have no direct control over this number and everytime we make a mistake like I [just did](https://toggl.slack.com/archives/C02APMYTM/p1570697984102400), there is no good way to increase/correct it. 

## The Solution
We want to solve this by parsing the release tags like this: 
```groovy
tag: android/ios-1.2.3-4 -> 
VERSION_MAJOR = 1 // max two digits
VERSION_MINOR = 2 // max two digits
VERSION_PATCH = 3 // max two digits
VERSION_BUILD = 4 // max three digits

// TODO we can also create even more room for bitrise build number  
// so we can distinguish adhoc builds as well

versionCode = VERSION_MAJOR * 10000000 + VERSION_MINOR * 100000 + VERSION_PATCH * 1000 + VERSION_BUILD
```
This should prevent situations like the [one I just created in release 2.4.1](https://toggl.slack.com/archives/C02APMYTM/p1570697984102400) and some other issues we had in the past. As a bonus, versionCode will have some semantic meaning now.

We don't need to make any change to our regular non-store builds.

### PR DESCRIPTION
**Pull request by semanticer**
## What's this?
This is a new way to increase versionCodes/bundleVersion for AppStore/PlayStore builds

### Relationships
Closes #6086

## Why do we want this?
We want to take more control over our version codes

## How is it done?
Parsing the tag name with regex ( thanks to @kornelijepetak and @juanlaube )

### Why not another way?
Show me

## Review hints
I don't even know how to test this properly :-) 

## Tests

Integration tests:  - do not merge this branch unless these tests are passing.

## :squid: Permissions
Feel free

### CODE REVIEW

File: build.cake [hunk 1]
```diff
@@ -93,6 +96,44 @@ private string GetCommitCount()
...
+    {
+        throw new InvalidOperationException($"Unable to get version number from this type of build target: {target}");
+    }
+    
+    IEnumerable<string> redirectedOutput;
+    StartProcess("git", new ProcessSettings
+    {
+        Arguments = "describe --tags",
+        RedirectStandardOutput = true
+    }, out redirectedOutput);
```
[the remark this answers is not in the document]
Comment (semanticer) [reply build.cake]: done

File: build.cake [hunk 2]
```diff
@@ -93,6 +96,44 @@ private string GetCommitCount()
     return redirectedOutput.Last();
 }
 
+private string GetVersionNumberFromTag()
+{
+    var platform = "";
+    if (target == "Build.Release.iOS.AppStore")
```
[the remark this answers is not in the document]
Comment (semanticer) [reply build.cake]: I thought we agreed we want to keep using commit count for `AdHoc` builds but maybe I just don't understand what `AdHoc` build exactly is :)

[the remark this answers is not in the document]
Comment (heytherewill) [reply build.cake]: No, that's fine, I was just wondering

File: build.cake [hunk 3]
```diff
@@ -93,6 +96,43 @@ private string GetCommitCount()
...
+         
+    var p = Regex.Match(tagName, @"(?<platform>(android|ios))-(?<major>\d{1,2})\.(?<minor>\d{1,2})\.(?<build>\d{1,2})(-(?<rev>\d{1,2}))?");
+    if (!p.Success) 
+    {
+        throw new InvalidOperationException($"Unsupported release tag format: {tagName}");
+    } 
+    var major = Int32.Parse(p.Groups["major"].Value) * 10000000;
+    var minor = Int32.Parse(p.Groups["minor"].Value) * 100000;
+    var build = Int32.Parse(p.Groups["build"].Value) * 1000;
+    var rev = string.IsNullOrEmpty(p.Groups["rev"].Value) ? Int32.Parse(p.Groups["rev"].Value) : 0;
```
Comment (amulware) [inline build.cake]: Please reduce the factors to 1_00_00_00, 1_00_00 and 1_00 - that's more than enough and it's what was agreed on internally.

File: build.cake [hunk 4]
```diff
@@ -93,6 +96,43 @@ private string GetCommitCount()
...
+    if (!p.Success) 
+    {
+        throw new InvalidOperationException($"Unsupported release tag format: {tagName}");
+    } 
+    var major = Int32.Parse(p.Groups["major"].Value) * 1000000;
+    var minor = Int32.Parse(p.Groups["minor"].Value) *   10000;
+    var build = Int32.Parse(p.Groups["build"].Value) *     100;
+    var rev = string.IsNullOrEmpty(p.Groups["rev"].Value) ? 0 : Int32.Parse(p.Groups["rev"].Value);
+
+    return (major + minor + build + rev).ToString();
```
Comment (kornelijepetak) [inline build.cake]: everything from here up until the return
```
var major = p.Groups["major"].Value;
...
...
...
return (major + minor + build + rev).ToString();
```

can be replaced by
```
return $"{p.Groups["major"]}{p.Groups["minor"],2}{p.Groups["build"],2}{p.Groups["rev"],2}".Replace(' ', '0');
```

Whether it's more readable, is up to you :)

Comment (semanticer) [reply build.cake]: Thanks for suggesting the alternative but I still prefer readability of the original solution and readability seems to be more important in this case.

**Pull request merged**